src/main/java/com/o19s/es/ltr/ranker/dectree/NaiveAdditiveDecisionTree.java

```diff
@@ -109,7 +121,7 @@ public float eval(float[] scores) {
             while (!n.isLeaf()) {
                 assert n instanceof Split;
                 Split s = (Split) n;
-                if (s.threshold > scores[s.feature]) {
+                if (comparer.compare(s.threshold, scores[s.feature])) {
```

Comment:
this wont work actually. I'm not sure if it actually uses it, but the way the lightgbm export is defined the comparer can be different for each split. We should use s.comparer.compare(...), although that might prevent inlining. I should run a benchmark to see if that matters at all.

Reply:
Coming from xgboost and not being very familiar with lightgbm- does lightgbm differentiate between `>=` and `>`? If not, would be possible in the case of `<`/`<=` to just swap left and right? It seems to make sense that all splits would be between buckets / sample values as opposed to on them. Reading below, I'm not sure how that might interact with the zero special case logic.

Related to the benchmark, the underlying GBT inference implementation here seems to be based on pointer-linked trees. At Yelp, we found pretty significant improvements from preordering the tree nodes in an array ([code](https://github.com/Yelp/xgboost-predictor-java/blob/master/src/main/java/biz/k11i/xgboost/tree/PreorderRegTree.java), [results](http://www.sysml.cc/doc/89.pdf)), but that definitely would be a different issue/PR.